src/main/kotlin/com/fasterxml/jackson/module/kotlin/ValueCreator.kt

```diff
@@ -0,0 +1,46 @@
+package com.fasterxml.jackson.module.kotlin
+
+import com.fasterxml.jackson.databind.DeserializationContext
+import com.fasterxml.jackson.databind.MapperFeature
+import kotlin.reflect.KFunction
+import kotlin.reflect.KParameter
+import kotlin.reflect.full.valueParameters
+
+/**
+ * A class that abstracts the creation of instances by calling KFunction.
+ * @see KotlinValueInstantiator
+ */
+internal sealed class ValueCreator<T> {
+    /**
+     * Function to be call.
+     */
+    protected abstract val callable: KFunction<T>
+
+    /**
+     * Initial value for accessibility by reflection.
+     */
+    protected abstract val accessible: Boolean
+
+    /**
+     * ValueParameters of the KFunction to be called.
+     */
+    val valueParameters: List<KParameter> by lazy { callable.valueParameters }
+
+    /**
+     * Checking process to see if access from context is possible.
+     * @throws  IllegalAccessException
+     */
+    fun checkAccessibility(ctxt: DeserializationContext) {
+        if ((!accessible && ctxt.config.isEnabled(MapperFeature.CAN_OVERRIDE_ACCESS_MODIFIERS)) ||
+            (accessible && ctxt.config.isEnabled(MapperFeature.OVERRIDE_PUBLIC_ACCESS_MODIFIERS))) {
+            return
+        }
+
+        throw IllegalAccessException("Cannot access to function or companion object instance, target: $callable")
```

Comment:
The error message here is a summary of what was originally individual content.
Should I follow the existing content and give a more specific message for each one?

Also, I couldn't think of any good error message content, so I would appreciate any suggestions.

Reply:
I'm not sure what you mean by individual content

Reply:
I don't remember exactly what I was thinking when I left this comment because it's been so long since I created the PR...

I think I was probably worrying about the following two points.

- Whether people who read this error message will be able to respond appropriately
- Should I indicate specifically which of `constructor`, `factory function`, or `companion object' was inaccessible?

Reply:
Ahh, yes, if that's possible I think it could prove to be quite helpful